Osc/170 implement a resource management component

lukas016 opened the pull request:
# Proposed Changes

Fixes #170

lukas016 committed:
draft: resource manager

lukas016 committed:
add unit tests for resource manager

lukas016 committed:
generalize allocate/deallocate functions

so-sahu commented on pkg/resources/manager/manager.go:
```diff
@@ -0,0 +1,561 @@
+	}
+
+	// dummy source uses reference for change available sources in runtime.
+	dummySource, ok := r.sources[sources.NewSourceDummy(nil).GetName()]
```
As discussed offline, we can delete this.

lukas016 replied:
fixed

so-sahu commented on pkg/resources/manager/manager.go:
```diff
@@ -0,0 +1,561 @@
+		if available.Value() < 0 {
+			return nil, ErrResourceNotAvailable
+		}
+		newAvailableResources[key] = *available
```
Use Sign method in place of value

lukas016 replied:
fixed

so-sahu commented on pkg/resources/manager/manager.go:
```diff
@@ -0,0 +1,561 @@
+			continue
+		}
+
+		quantity.Add(resource)
```
Please add some suitable comment here.

lukas016 replied:
fixed

lukas016 commented on pkg/controllers/machine_controller.go:
```diff
@@ -352,6 +357,12 @@ func (r *MachineReconciler) processMachineDeletion(ctx context.Context, log logr
 	}
 	log.V(1).Info("Removed Finalizer. Deletion completed")
 
+	err = manager.Deallocate(machine, machine.Spec.Resources.DeepCopy())
```
wrong place for this function, it has to be before removing of finalizers

lukas016 replied:
fixed

lukas016 commented on pkg/resources/manager/manager.go:
```diff
@@ -0,0 +1,561 @@
+// reset internal state of manager and allow reinit
+// Use it for unit test only.
+func (r *resourceManager) reset() {
+	mng.ctx = nil
```
i checked logic again and this solution is better. Main problem with your suggestion is reference values as sources, or available resources. Their default value is nil. So if i create only new object, it contains nil and i will have to extend logic in another functions for this specific case. Reset function solving this problem for me because it initialize reference types.

lukas016 committed:
apply review comments

- fix wrong using of global env
- improve deallocate of resources for vm deletion
- delete letfover logic for dummy source
- use Sign method for check allocation result
- add comments

so-sahu reviewed (approved):
LGTM.
Integration tests are not yet compatible with the resource manager. However, as we discussed offline, this will be addressed later.
